astropy/io/ascii/tests/test_arg_write.py

```diff
@@ -0,0 +1,43 @@
+import warnings
+
+from ... import ascii
+from ....tests.helper import pytest, catch_warnings
+from ....utils.exceptions import AstropyWarning
+from ....extern.six.moves import cStringIO as StringIO
+
+_data = ([1,2,3],[4.1,5.2,6.3])
+
+'''
+**********************************************************************************
+** i set the test cases in base of output if output value is true means warning **
+** accure otherwise output value false                                          **
+**********************************************************************************
+'''
+
+'''
+it's check one condition ::
+(i) when names keys mismatch formats keys
+If any condition true then it's make warning
+Example ::
+>>> from astropy.io import ascii
+>>> data = ([1],[4],['hello'])
+>>> ascii.write(data, names=['x', 'y', 'z'], formats={'p':'%d', 'y':'%d'})
+WARNING: {'p'} keys in the formats not to match the keys in names [astropy.io.ascii.ui]
+x y z
+1 4 hello
+'''
+
+@pytest.mark.parametrize('names, formats, output', [
+    (['x', 'y'], {'p':'%d', 'y':'%f'}, True),
+    (['x', 'y'], {'p':'%d', 'q':'%f'}, True),
+    (['x', 'y'], {'x':'%d', 'y':'%f'}, False),
+    (['x', 'y'], {'z':'%f'}, True),
+    (['x', 'y'], {'x':'%d'}, False),
+    (['x', 'y'], {'y':'%f'}, False),
+    (['x', 'y'], {}, False)
+])
+def test_name_with_formats(names, formats, output):
+    with catch_warnings(AstropyWarning) as ASwarn:
+        out = StringIO()
+        ascii.write(_data, out, names=names, formats=formats)
```

Comment:
This doesn't test the most common case that the table (`_data` here) is defining the column names.